File: integrations/anythingllm/model-neutral-compatibility/visual_atlas.py
#!/usr/bin/env python3
"""Canonical Visual Atlas catalog and deterministic selective retrieval.

The manifest remains the authority for entries and assets. This module only
adds a small, reviewed alias layer and never performs inference or generation.
"""

from __future__ import annotations

import hashlib
import json
import os
import re
import unicodedata
from dataclasses import dataclass
from pathlib import Path
from typing import Any


ATLAS_MODULE = "visual-atlas"
PLAN_SCHEMA = "aag.selective-knowledge.plan.v1"
CATALOG_SCHEMA = "aag.visual-style-atlas.catalog.v1"
SUPPORTED_MODES = {"auto", "manual_taxonomy", "manual_browse"}
SAFE_ID_RE = re.compile(r"^[a-z0-9]+(?:-[a-z0-9]+)*$")


class AtlasError(ValueError):
    pass


def _project_root() -> Path:
    override = os.environ.get("AAG_IMAGE_PROJECT_ROOT")
    if override:
        return Path(override).resolve()
    return Path(__file__).resolve().parents[3]


def _atlas_root() -> Path:
    override = os.environ.get("AAG_VISUAL_ATLAS_ROOT")
    if override:
        return Path(override).resolve()
    return _project_root() / "visual-atlas"


def _json(path: Path) -> dict[str, Any]:
    try:
        value = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError) as error:
        raise AtlasError(f"Visual Atlas data is unavailable or invalid: {path.name}") from error
    if not isinstance(value, dict):
        raise AtlasError(f"Visual Atlas data is not an object: {path.name}")
    return value


def _sha256(path: Path) -> str:
    digest = hashlib.sha256()
    with path.open("rb") as handle:
        for block in iter(lambda: handle.read(1024 * 1024), b""):
            digest.update(block)
    return digest.hexdigest()


def normalize_text(value: Any) -> str:
    text = unicodedata.normalize("NFKC", str(value or "")).casefold()
    text = text.replace("’", "'").replace("־", "-")
    text = re.sub(r"[^\w\u0590-\u05ff]+", " ", text, flags=re.UNICODE)
    return re.sub(r"\s+", " ", text).strip()


def _contains(text: str, phrase: str) -> bool:
    return bool(phrase) and f" {phrase} " in f" {text} "


@dataclass(frozen=True)
class AtlasPaths:
    taxonomy: Path
    manifest: Path
    aliases: Path


class VisualAtlas:
    """Validated, immutable view over the completed Atlas."""

    def __init__(self, paths: AtlasPaths | None = None, *, verify_assets: bool = False):
        atlas_root = _atlas_root()
        self.paths = paths or AtlasPaths(
            taxonomy=Path(__file__).resolve().parent / "composer" / "visual-taxonomy.json",
            manifest=atlas_root / "manifest" / "atlas-manifest.json",
            aliases=atlas_root / "manifest" / "retrieval-aliases.json",
        )
        self.taxonomy = _json(self.paths.taxonomy)
        self.manifest = _json(self.paths.manifest)
        self.alias_config = _json(self.paths.aliases)
        self.atlas_root = self.paths.manifest.parent.parent.resolve()
        product = _json(self.atlas_root / "manifest" / "product-assets.json")
        self.product_assets = {item.get("key"): item for item in product.get("entries", [])}
        self._validate(verify_assets=verify_assets)

    def _validate(self, *, verify_assets: bool) -> None:
        taxonomy_sha = _sha256(self.paths.taxonomy)
        if self.manifest.get("taxonomy_sha256") != taxonomy_sha:
            raise AtlasError("Visual Atlas taxonomy hash does not match the canonical manifest.")
        if self.alias_config.get("atlas_version") != self.manifest.get("atlas_version"):
            raise AtlasError("Visual Atlas alias metadata targets a different Atlas version.")

        pairs: list[tuple[str, str]] = []
        self.family_labels: dict[str, str] = {}
        self.subfamily_labels: dict[tuple[str, str], str] = {}
        for family in self.taxonomy.get("families", []):
            family_id = family.get("id")
            if not isinstance(family_id, str) or not SAFE_ID_RE.fullmatch(family_id):
                raise AtlasError("Visual Atlas taxonomy contains an invalid family ID.")
            self.family_labels[family_id] = str(family.get("label") or family_id)
            for subfamily in family.get("subfamilies", []):
                subfamily_id = subfamily.get("id")
                if not isinstance(subfamily_id, str) or not SAFE_ID_RE.fullmatch(subfamily_id):
                    raise AtlasError("Visual Atlas taxonomy contains an invalid subfamily ID.")
                pair = (family_id, subfamily_id)
                if pair in self.subfamily_labels:
                    raise AtlasError("Visual Atlas taxonomy contains a duplicate style.")
                pairs.append(pair)
                self.subfamily_labels[pair] = str(subfamily.get("label") or subfamily_id)

        entries = self.manifest.get("entries")
        if not isinstance(entries, list) or len(entries) != 493 or len(pairs) != 493:
            raise AtlasError("Visual Atlas must contain exactly 493 canonical entries.")
        self.entries: dict[tuple[str, str], dict[str, Any]] = {}
        self.thumbnail_sha256: dict[tuple[str, str], str] = {}
        for entry in entries:
            pair = (entry.get("family_id"), entry.get("subfamily_id"))
            if pair not in self.subfamily_labels or pair in self.entries:
                raise AtlasError("Visual Atlas manifest does not map one-to-one to the taxonomy.")
            if entry.get("status") != "COMPLETED":
                raise AtlasError("Visual Atlas contains an entry that is not completed.")
            output = self._asset_path(entry.get("output_path"), ".png")
            thumbnail = self._asset_path(entry.get("thumbnail_path"), ".webp")
            output_present = output.is_file()
            thumbnail_present = thumbnail.is_file()
            if output_present != thumbnail_present:
                raise AtlasError("Visual Atlas contains a partial preview/thumbnail pair.")
            if verify_assets and not output_present:
                raise AtlasError("Visual Atlas pixel verification was requested but assets are missing.")
            if output_present and (output.stat().st_size < 128 or thumbnail.stat().st_size < 128):
                raise AtlasError("Visual Atlas contains an empty preview or thumbnail.")
            if output_present and thumbnail.read_bytes()[:4] != b"RIFF":
                raise AtlasError("Visual Atlas contains an invalid thumbnail header.")
            if verify_assets and _sha256(output) != entry.get("sha256"):
                raise AtlasError("Visual Atlas preview hash validation failed.")
            self.entries[pair] = {**entry, "assets_available": output_present}
            # Thumbnail derivatives have their own immutable browser cache
            # identity. Tying their URL to the reference PNG hash leaves a
            # repaired derivative trapped behind its previous cached bytes.
            product_entry = self.product_assets.get(f"{pair[0]}/{pair[1]}", {})
            recorded_thumbnail = product_entry.get("thumbnail", {}).get("sha256")
            if not thumbnail_present and not isinstance(recorded_thumbnail, str):
                raise AtlasError("Visual Atlas metadata lacks the thumbnail integrity record.")
            self.thumbnail_sha256[pair] = _sha256(thumbnail) if thumbnail_present else recorded_thumbnail
        if set(pairs) != set(self.entries):
            raise AtlasError("Visual Atlas manifest and taxonomy style sets differ.")

        raw_aliases = self.alias_config.get("entries", {})
        if not isinstance(raw_aliases, dict):
            raise AtlasError("Visual Atlas aliases are invalid.")
        self.aliases: dict[tuple[str, str], list[str]] = {}
        for key, values in raw_aliases.items():
            bits = str(key).split("/", 1)
            pair = tuple(bits) if len(bits) == 2 else ("", "")
            if pair not in self.entries or not isinstance(values, list) or not all(isinstance(value, str) for value in values):
                raise AtlasError(f"Visual Atlas alias target is invalid: {key}")
            normalized = sorted({normalize_text(value) for value in values if normalize_text(value)}, key=lambda item: (-len(item.split()), item))
            self.aliases[pair] = normalized
        self.style_cues = tuple(normalize_text(value) for value in self.alias_config.get("style_cues", []) if normalize_text(value))
        self.top_k = min(2, max(1, int(self.alias_config.get("top_k", 2))))
        self.minimum_confidence = float(self.alias_config.get("minimum_confidence", 0.72))
        self.taxonomy_sha256 = taxonomy_sha
        self.manifest_sha256 = _sha256(self.paths.manifest)

    def _asset_path(self, relative: Any, suffix: str) -> Path:
        if not isinstance(relative, str) or not relative.endswith(suffix) or relative.startswith(("/", "~")):
            raise AtlasError("Visual Atlas contains an unsafe asset path.")
        resolved = (self.atlas_root / relative).resolve()
        if self.atlas_root not in resolved.parents:
            raise AtlasError("Visual Atlas asset path escapes its root.")
        return resolved

    def entry(self, family_id: str, subfamily_id: str) -> dict[str, Any]:
        try:
            return self.entries[(family_id, subfamily_id)]
        except KeyError as error:
            raise AtlasError("Unknown Visual Atlas family/subfamily selection.") from error

    def asset(self, family_id: str, subfamily_id: str, kind: str) -> tuple[Path, str]:
        entry = self.entry(family_id, subfamily_id)
        if not entry.get("assets_available"):
            raise AtlasError("Visual Atlas pixels are not installed; metadata-only mode is active.")
        if kind == "thumbnail":
            return self._asset_path(entry["thumbnail_path"], ".webp"), "image/webp"
        if kind == "preview":
            return self._asset_path(entry["output_path"], ".png"), "image/png"
        raise AtlasError("Unknown Visual Atlas asset kind.")

    def _public_entry(self, pair: tuple[str, str]) -> dict[str, Any]:
        entry = self.entries[pair]
        family_id, subfamily_id = pair
        return {
            # Retain the canonical taxonomy shape consumed by every existing
            # Composer selector while adding manifest-backed Atlas metadata.
            "id": subfamily_id,
            "label": self.subfamily_labels[pair],
            "index": entry["index"],
            "family_id": family_id,
            "family_label": self.family_labels[family_id],
            "subfamily_id": subfamily_id,
            "subfamily_label": self.subfamily_labels[pair],
            "description": entry["style_descriptor"],
            "aliases": self.aliases.get(pair, []),
            "atlas": {
                "available": bool(entry.get("assets_available")),
                "sha256": entry["sha256"],
                "thumbnail_sha256": self.thumbnail_sha256[pair],
                "width": entry["width"],
                "height": entry["height"],
            },
        }

    def catalog(self) -> dict[str, Any]:
        families = []
        for family in self.taxonomy["families"]:
            family_id = family["id"]
            families.append({
                "id": family_id,
                "label": self.family_labels[family_id],
                "classification": family.get("classification"),
                "subfamily_classification": family.get("subfamily_classification"),
                "subfamilies": [self._public_entry((family_id, subfamily["id"])) for subfamily in family["subfamilies"]],
            })
        return {
            "schema": CATALOG_SCHEMA,
            "atlas_version": self.manifest["atlas_version"],
            "taxonomy_sha256": self.taxonomy_sha256,
            "manifest_sha256": self.manifest_sha256,
            "total_entries": len(self.entries),
            "families": families,
        }

    def _selection(self, pair: tuple[str, str], *, confidence: float, matched: list[str]) -> dict[str, Any]:
        entry = self.entries[pair]
        return {
            "family_id": pair[0],
            "family_label": self.family_labels[pair[0]],
            "subfamily_id": pair[1],
            "subfamily_label": self.subfamily_labels[pair],
            "atlas_index": entry["index"],
            "style_descriptor": entry["style_descriptor"],
            "confidence": round(confidence, 3),
            "matched": matched[:4],
        }

    def _plan(self, *, used: bool, mode: str, reason: str, selections: list[dict[str, Any]] | None = None) -> dict[str, Any]:
        selected = selections or []
        return {
            "schema": PLAN_SCHEMA,
            "module": ATLAS_MODULE,
            "used": used,
            "mode": mode,
            "reason": reason,
            "confidence": max((item["confidence"] for item in selected), default=0.0),
            "selections": selected[: self.top_k],
            "visual_reference_used": False,
            "context_chars": 0,
            "estimated_context_tokens": 0,
            "taxonomy_sha256": self.taxonomy_sha256,
            "manifest_sha256": self.manifest_sha256,
        }

    def select(
        self,
        text: Any,
        *,
        mode: str = "auto",
        family_id: str | None = None,
        subfamily_id: str | None = None,
        operation: str = "generate",
        preservation: str = "none",
    ) -> dict[str, Any]:
        if mode not in SUPPORTED_MODES:
            raise AtlasError("Unknown Visual Atlas selection mode.")
        if operation == "upscale":
            return self._plan(used=False, mode=mode, reason="operation_excluded")
        if preservation == "identity":
            return self._plan(used=False, mode=mode, reason="identity_reference_protected")

        if mode != "auto" or (family_id and subfamily_id):
            if not family_id or not subfamily_id:
                raise AtlasError("Manual Visual Atlas selection requires a family and subfamily.")
            pair = (family_id, subfamily_id)
            self.entry(*pair)
            return self._plan(
                used=True,
                mode=mode if mode != "auto" else "manual_taxonomy",
                reason="explicit_user_selection",
                selections=[self._selection(pair, confidence=1.0, matched=["manual"])],
            )

        normalized = normalize_text(text)
        if not normalized:
            return self._plan(used=False, mode="auto", reason="empty_request")
        has_style_cue = any(_contains(normalized, cue) for cue in self.style_cues)
        if not has_style_cue:
            return self._plan(used=False, mode="auto", reason="no_style_intent")

        scores: dict[tuple[str, str], tuple[float, list[str]]] = {}
        for pair, aliases in self.aliases.items():
            matched = [alias for alias in aliases if _contains(normalized, alias)]
            if matched:
                longest = max(len(alias.split()) for alias in matched)
                scores[pair] = (min(0.99, 0.88 + 0.02 * longest), matched)

        # Canonical labels are useful fallback aliases, but ties remain unresolved.
        for pair in self.entries:
            family_phrase = normalize_text(self.family_labels[pair[0]])
            subfamily_phrase = normalize_text(self.subfamily_labels[pair])
            if _contains(normalized, subfamily_phrase) and pair not in scores:
                words = len(subfamily_phrase.split())
                confidence = 0.75 if words >= 2 or _contains(normalized, family_phrase) else 0.72
                scores[pair] = (confidence, [subfamily_phrase])

        ranked = sorted(scores.items(), key=lambda item: (-item[1][0], self.entries[item[0]]["index"]))
        if not ranked or ranked[0][1][0] < self.minimum_confidence:
            return self._plan(used=False, mode="auto", reason="no_reliable_match")
        if len(ranked) > 1 and ranked[0][1][0] == ranked[1][1][0] and ranked[0][1][0] < 0.88:
            return self._plan(used=False, mode="auto", reason="ambiguous_match")

        top_score = ranked[0][1][0]
        chosen = []
        for pair, (score, matched) in ranked:
            if len(chosen) >= self.top_k or score < max(self.minimum_confidence, top_score - 0.08):
                break
            chosen.append(self._selection(pair, confidence=score, matched=matched))
        return self._plan(used=True, mode="auto", reason="deterministic_alias_match", selections=chosen)


_INSTANCE: VisualAtlas | None = None


def get_visual_atlas() -> VisualAtlas:
    global _INSTANCE
    if _INSTANCE is None:
        _INSTANCE = VisualAtlas()
    return _INSTANCE

File: image-system/src/visual-atlas.js
"use strict";

const fs = require("fs");
const path = require("path");
const crypto = require("crypto");
const { AagError } = require("./errors");

const MODULE = "visual-atlas";
const PLAN_SCHEMA = "aag.selective-knowledge.plan.v1";
const MARKER_PREFIX = "AAG_ATLAS_SELECTION_V1";
const SAFE_ID = /^[a-z0-9]+(?:-[a-z0-9]+)*$/;
const MODES = new Set(["auto", "manual_taxonomy", "manual_browse"]);
const MAX_CONTEXT_CHARS = 720;
let cached = null;

function sha256File(file) {
  return crypto.createHash("sha256").update(fs.readFileSync(file)).digest("hex");
}

function normalizeText(value) {
  return String(value || "").normalize("NFKC").toLocaleLowerCase("und")
    .replace(/[’]/gu, "'").replace(/[־]/gu, "-")
    .replace(/[^\p{L}\p{N}_\u0590-\u05ff]+/gu, " ").replace(/\s+/gu, " ").trim();
}

function contains(text, phrase) {
  return Boolean(phrase) && ` ${text} `.includes(` ${phrase} `);
}

function resolveAtlasRoot() {
  const candidates = [
    process.env.AAG_VISUAL_ATLAS_ROOT,
    "/app/server/storage/aag-visual-atlas",
    path.resolve(__dirname, "../..", "visual-atlas"),
  ].filter(Boolean);
  const root = candidates.find(candidate => fs.existsSync(path.join(candidate, "manifest", "atlas-manifest.json")));
  if (!root) throw new AagError("ATLAS_UNAVAILABLE", "The Visual Atlas manifest is unavailable.");
  return fs.realpathSync(root);
}

function readJson(file) {
  try { return JSON.parse(fs.readFileSync(file, "utf8")); }
  catch (error) { throw new AagError("ATLAS_INVALID", "The Visual Atlas metadata is invalid.", false, error?.message); }
}

function safeAsset(root, relative, suffix) {
  if (typeof relative !== "string" || !relative.endsWith(suffix) || path.isAbsolute(relative)) throw new AagError("ATLAS_INVALID", "The Visual Atlas contains an unsafe asset path.");
  const resolved = path.resolve(root, relative);
  if (!resolved.startsWith(`${root}${path.sep}`)) throw new AagError("ATLAS_INVALID", "The Visual Atlas asset escapes its root.");
  return resolved;
}

function load() {
  if (cached) return cached;
  const root = resolveAtlasRoot();
  const manifestFile = path.join(root, "manifest", "atlas-manifest.json");
  const aliasFile = path.join(root, "manifest", "retrieval-aliases.json");
  const taxonomyCandidates = [
    process.env.AAG_VISUAL_TAXONOMY_PATH,
    path.join(__dirname, "visual-taxonomy.json"),
    path.join(root, "..", "image-agent", "integrations", "model-neutral-compatibility", "composer", "visual-taxonomy.json"),
    path.resolve(__dirname, "..", "integrations", "model-neutral-compatibility", "composer", "visual-taxonomy.json"),
  ].filter(Boolean);
  const taxonomyFile = taxonomyCandidates.find(file => fs.existsSync(file));
  if (!taxonomyFile) throw new AagError("ATLAS_UNAVAILABLE", "The canonical Visual Atlas taxonomy is unavailable.");
  const manifest = readJson(manifestFile);
  const aliases = readJson(aliasFile);
  const taxonomy = readJson(taxonomyFile);
  const taxonomySha256 = sha256File(taxonomyFile);
  if (manifest.taxonomy_sha256 !== taxonomySha256) throw new AagError("ATLAS_INVALID", "The Visual Atlas taxonomy hash does not match its manifest.");
  if (aliases.atlas_version !== manifest.atlas_version) throw new AagError("ATLAS_INVALID", "The Visual Atlas alias layer targets a different Atlas version.");

  const familyLabels = new Map();
  const subfamilyLabels = new Map();
  const canonicalPairs = [];
  for (const family of taxonomy.families || []) {
    if (!SAFE_ID.test(family.id)) throw new AagError("ATLAS_INVALID", "The Visual Atlas contains an invalid family ID.");
    familyLabels.set(family.id, String(family.label || family.id));
    for (const subfamily of family.subfamilies || []) {
      if (!SAFE_ID.test(subfamily.id)) throw new AagError("ATLAS_INVALID", "The Visual Atlas contains an invalid subfamily ID.");
      const key = `${family.id}/${subfamily.id}`;
      if (subfamilyLabels.has(key)) throw new AagError("ATLAS_INVALID", "The Visual Atlas contains a duplicate style.");
      subfamilyLabels.set(key, String(subfamily.label || subfamily.id));
      canonicalPairs.push(key);
    }
  }
  if (canonicalPairs.length !== 493 || !Array.isArray(manifest.entries) || manifest.entries.length !== 493) throw new AagError("ATLAS_INVALID", "The completed Visual Atlas must contain exactly 493 styles.");
  const entries = new Map();
  for (const entry of manifest.entries) {
    const key = `${entry.family_id}/${entry.subfamily_id}`;
    if (!subfamilyLabels.has(key) || entries.has(key) || entry.status !== "COMPLETED") throw new AagError("ATLAS_INVALID", "The Visual Atlas manifest does not map one-to-one to completed taxonomy entries.");
    const preview = safeAsset(root, entry.output_path, ".png");
    const thumbnail = safeAsset(root, entry.thumbnail_path, ".webp");
    const previewPresent = fs.existsSync(preview);
    const thumbnailPresent = fs.existsSync(thumbnail);
    if (previewPresent !== thumbnailPresent) throw new AagError("ATLAS_INVALID", "The Visual Atlas contains a partial asset pair.");
    if (previewPresent && (!fs.statSync(preview).isFile() || !fs.statSync(thumbnail).isFile() || fs.statSync(preview).size < 128 || fs.statSync(thumbnail).size < 128 || fs.readFileSync(thumbnail).subarray(0, 4).toString("ascii") !== "RIFF")) throw new AagError("ATLAS_INVALID", "The Visual Atlas contains an empty or invalid asset.");
    entries.set(key, { ...entry, assets_available: previewPresent });
  }
  if (canonicalPairs.some(key => !entries.has(key))) throw new AagError("ATLAS_INVALID", "The Visual Atlas taxonomy and manifest style sets differ.");

  const aliasMap = new Map();
  for (const [key, values] of Object.entries(aliases.entries || {})) {
    if (!entries.has(key) || !Array.isArray(values) || values.some(value => typeof value !== "string")) throw new AagError("ATLAS_INVALID", `The Visual Atlas alias target is invalid: ${key}`);
    aliasMap.set(key, [...new Set(values.map(normalizeText).filter(Boolean))].sort((a, b) => b.split(" ").length - a.split(" ").length || a.localeCompare(b)));
  }
  cached = Object.freeze({
    root, manifest, taxonomy, aliases, entries, familyLabels, subfamilyLabels, aliasMap,
    styleCues: (aliases.style_cues || []).map(normalizeText).filter(Boolean),
    topK: Math.min(2, Math.max(1, Number(aliases.top_k || 2))),
    minimumConfidence: Number(aliases.minimum_confidence || 0.72),
    taxonomySha256,
    manifestSha256: sha256File(manifestFile),
  });
  return cached;
}

function selection(data, key, confidence, matched) {
  const entry = data.entries.get(key);
  const [familyId, subfamilyId] = key.split("/");
  return {
    family_id: familyId,
    family_label: data.familyLabels.get(familyId),
    subfamily_id: subfamilyId,
    subfamily_label: data.subfamilyLabels.get(key),
    atlas_index: entry.index,
    style_descriptor: entry.style_descriptor,
    confidence: Number(confidence.toFixed(3)),
    matched: matched.slice(0, 4),
  };
}

function plan(data, { used = false, mode = "auto", reason, selections = [] }) {
  return {
    schema: PLAN_SCHEMA,
    module: MODULE,
    used,
    mode,
    reason,
    confidence: selections.reduce((best, item) => Math.max(best, item.confidence), 0),
    selections: selections.slice(0, data?.topK || 2),
    visual_reference_used: false,
    context_chars: 0,
    estimated_context_tokens: 0,
    taxonomy_sha256: data?.taxonomySha256 || null,
    manifest_sha256: data?.manifestSha256 || null,
  };
}

function select(text, options = {}) {
  const data = load();
  const mode = options.mode || "auto";
  if (!MODES.has(mode)) throw new AagError("ATLAS_SELECTION_INVALID", "The Visual Atlas selection mode is invalid.");
  if (options.operation === "upscale") return plan(data, { mode, reason: "operation_excluded" });
  if (options.preservation === "identity") return plan(data, { mode, reason: "identity_reference_protected" });

  const familyId = options.familyId || null;
  const subfamilyId = options.subfamilyId || null;
  if (mode !== "auto" || (familyId && subfamilyId)) {
    if (!familyId || !subfamilyId) throw new AagError("ATLAS_SELECTION_INVALID", "Manual Visual Atlas selection requires a family and subfamily.");
    const key = `${familyId}/${subfamilyId}`;
    if (!data.entries.has(key)) throw new AagError("ATLAS_SELECTION_INVALID", "The selected Visual Atlas style does not exist.");
    return plan(data, { used: true, mode: mode === "auto" ? "manual_taxonomy" : mode, reason: "explicit_user_selection", selections: [selection(data, key, 1, ["manual"])] });
  }

  const normalized = normalizeText(text);
  if (!normalized) return plan(data, { mode: "auto", reason: "empty_request" });
  if (!data.styleCues.some(cue => contains(normalized, cue))) return plan(data, { mode: "auto", reason: "no_style_intent" });
  const scores = new Map();
  for (const [key, aliases] of data.aliasMap) {
    const matched = aliases.filter(alias => contains(normalized, alias));
    if (matched.length) {
      const longest = Math.max(...matched.map(alias => alias.split(" ").length));
      scores.set(key, [Math.min(0.99, 0.88 + 0.02 * longest), matched]);
    }
  }
  for (const key of data.entries.keys()) {
    if (scores.has(key)) continue;
    const [familyId] = key.split("/");
    const familyPhrase = normalizeText(data.familyLabels.get(familyId));
    const subfamilyPhrase = normalizeText(data.subfamilyLabels.get(key));
    if (contains(normalized, subfamilyPhrase)) {
      const confidence = subfamilyPhrase.split(" ").length >= 2 || contains(normalized, familyPhrase) ? 0.75 : 0.72;
      scores.set(key, [confidence, [subfamilyPhrase]]);
    }
  }
  const ranked = [...scores.entries()].sort((a, b) => b[1][0] - a[1][0] || data.entries.get(a[0]).index - data.entries.get(b[0]).index);
  if (!ranked.length || ranked[0][1][0] < data.minimumConfidence) return plan(data, { mode: "auto", reason: "no_reliable_match" });
  if (ranked.length > 1 && ranked[0][1][0] === ranked[1][1][0] && ranked[0][1][0] < 0.88) return plan(data, { mode: "auto", reason: "ambiguous_match" });
  const best = ranked[0][1][0];
  const chosen = ranked.filter(([, [score]]) => score >= Math.max(data.minimumConfidence, best - 0.08)).slice(0, data.topK)
    .map(([key, [score, matched]]) => selection(data, key, score, matched));
  return plan(data, { used: true, mode: "auto", reason: "deterministic_alias_match", selections: chosen });
}

function marker(mode, familyId, subfamilyId) {
  if (!MODES.has(mode) || mode === "auto" || !SAFE_ID.test(familyId || "") || !SAFE_ID.test(subfamilyId || "")) throw new AagError("ATLAS_SELECTION_INVALID", "The Visual Atlas marker fields are invalid.");
  return `${MARKER_PREFIX} mode=${mode} family=${familyId} subfamily=${subfamilyId}`;
}

function parseMarker(value) {
  const match = String(value || "").match(/(?:^|\n)AAG_ATLAS_SELECTION_V1 mode=(manual_taxonomy|manual_browse) family=([a-z0-9]+(?:-[a-z0-9]+)*) subfamily=([a-z0-9]+(?:-[a-z0-9]+)*)(?:\n|$)/u);
  return match ? { mode: match[1], familyId: match[2], subfamilyId: match[3] } : null;
}

function contextFor(selected) {
  if (!selected?.length) return "";
  const styles = selected.map(item => `${item.family_label} → ${item.subfamily_label}: ${item.style_descriptor}`).join(" ");
  return `Visual Atlas style guidance: ${styles} Apply only the visual presentation, medium, composition, lighting, and rendering traits. Keep the user's subject and requested details authoritative; do not copy the Atlas benchmark subject. Human anatomy and identity-preservation constraints remain authoritative.`.slice(0, MAX_CONTEXT_CHARS);
}

function applyToTask(task, { logger } = {}) {
  const explicit = parseMarker(task?._aag_upstream_request);
  let atlasPlan;
  try {
    atlasPlan = select(task?._aag_authoritative_request || task?.request || task?.prompt, {
      mode: explicit?.mode || "auto",
      familyId: explicit?.familyId,
      subfamilyId: explicit?.subfamilyId,
      operation: task?.operation,
      preservation: task?.preservation,
    });
  } catch (error) {
    if (explicit) throw error;
    logger?.(`[AAG-IMAGE] Visual Atlas auto retrieval bypassed: ${String(error?.code || error?.message || "unavailable")}`);
    atlasPlan = plan(null, { mode: "auto", reason: "atlas_unavailable" });
  }
  if (atlasPlan.used) {
    const context = contextFor(atlasPlan.selections);
    task.prompt = `${String(task.prompt || "").trim()}\n\n${context}`;
    atlasPlan.context_chars = context.length;
    atlasPlan.estimated_context_tokens = Math.ceil(Buffer.byteLength(context, "utf8") / 4);
    if (task._aag_prompt_contract) {
      task._aag_prompt_contract = {
        ...task._aag_prompt_contract,
        selective_knowledge: [MODULE],
        pre_knowledge_prompt_sha256: task._aag_prompt_contract.final_prompt_sha256,
        final_prompt_sha256: crypto.createHash("sha256").update(task.prompt).digest("hex"),
      };
    }
  }
  task.atlas = atlasPlan;
  return task;
}

function resetCacheForTests() { cached = null; }

module.exports = { MODULE, PLAN_SCHEMA, MARKER_PREFIX, MAX_CONTEXT_CHARS, normalizeText, load, select, marker, parseMarker, contextFor, applyToTask, resetCacheForTests };
